x-pack/plugins/ml/common/util/job_utils.ts

```diff
@@ -85,6 +73,29 @@ export function isMappableJob(job: CombinedJob, detectorIndex: number): boolean
   return isMappable;
 }
 
+export function hasValidComposite(buckets: { [key: string]: any }) {
+  if (
+    isPopulatedObject(buckets, ['composite']) &&
+    isPopulatedObject(buckets.composite, ['sources'])
+  ) {
+    const sources = buckets.composite.sources as Array<Record<string, object>>;
+    return sources
+      .map((source) => {
+        const sourceName = getFirstKeyInObject(source);
+        if (sourceName !== undefined && isPopulatedObject(source[sourceName])) {
+          const sourceTypes = Object.keys(source[sourceName]);
+          return (
+            sourceTypes.length === 1 &&
+            (sourceTypes[0] === 'date_histogram' || sourceTypes[0] === 'terms')
+          );
+        }
+        return false;
+      })
+      .every((isValidCompositeSource: boolean) => isValidCompositeSource === true);
```

Comment:
I assume there is no need in using `map`, and also no need to check that every composite source is valid. If at least one is invalid, you already know the result.  Consider checking it with `some` instead.

Reply:
Updated here [`25af35d` (#96348)](https://github.com/elastic/kibana/pull/96348/commits/25af35d0b966d98edd92a726b6fb74954a3ccc7c)